COMM-455 update to net472

JasonBridgman-Nearmap opened the pull request.

JasonBridgman-Nearmap committed:
Updated to .Net 4.7.2

JasonBridgman-Nearmap committed:
Updated Autofixture Nuget to 4.17.0

JasonBridgman-Nearmap committed:
Updated Automapper to 10.1.1

JasonBridgman-Nearmap committed:
Updated Castle.Core to 4.4.1

JasonBridgman-Nearmap committed:
Updated Logging Nugets
- Common.Logging to 3.4.1
- Common.Logging.Core to 3.4.1
- Common.Logging.Serilog to 4.4.0

JasonBridgman-Nearmap committed:
Removed obsolete Nugets
- Liblog 4.2.5
- NHibernate.Logging 4.0.3
- Serilog.Sinks.RollingFile 2.1.0

JasonBridgman-Nearmap committed:
Updated IdentityServer3 Nuget to 2.6.3

JasonBridgman-Nearmap committed:
Fixed Analysis Warnings

JasonBridgman-Nearmap committed:
Updated Microsoft compiler nugets

JasonBridgman-Nearmap committed:
Fixed issues
Fixed issues preventing Web project from starting, added Postgres driver. (only required if using postgres db).
Refactored postgres support classes.

JasonBridgman-Nearmap committed:
Updated Microsoft.Owin nugets to 4.2.0

JasonBridgman-Nearmap committed:
Updated Moq nuget to 4.17.0

JasonBridgman-Nearmap committed:
Updated Newtonsoft.Json nuget to 13.0.1

JasonBridgman-Nearmap committed:
Removed psake

JasonBridgman-Nearmap committed:
Updated Serilog nugets
- Serilog to 2.10.0
- Serilog.Sinks.File to 5.0.0
- Serilog.Sinks.Trace to 3.0.0

JasonBridgman-Nearmap committed:
Updated nuget package Fare to 2.2.0

JasonBridgman-Nearmap committed:
Updated System nugets
- System.Text.Json 6.0.2
- System.Runtime.CompilerServices.Unsafe 6.0.0

JasonBridgman-Nearmap committed:
Updated xUnit nugets
- xunit 2.4.1
- xunit.abstractions 2.0.3
- xunit.analyzers 0.10.0
- xunit.assert 2.4.1
- xunit.core 2.4.1
- xunit.extensibility.core 2.4.1
- xunit.extensibility.execution 2.4.1
- xunit.runner.console 2.4.1
- xunit.runner.visualstudio 2.4.3

JasonBridgman-Nearmap committed:
Updated nuget FluentNHibernate to 3.1.0

JasonBridgman-Nearmap committed:
Removed LibLog files.
Removed LibLog package files and replaced with Microsoft.Extensions.Logging.Abstractions nuget package v6.0.1

JasonBridgman-Nearmap committed:
Bumped Assembly and assembly file versions to 2.

JasonBridgman-Nearmap committed:
Refactor
Removed unused usings, blank lines and restructured some methods.

JasonBridgman-Nearmap committed:
Removed AsyncHelper

JasonBridgman-Nearmap committed:
Removed ScopeExtensions

JasonBridgman-Nearmap committed:
Refactored unit tests

JasonBridgman-Nearmap committed:
Refactored IdentityServerServiceFactoryExtensions

JasonBridgman-Nearmap committed:
Removed ClientExtensions

JasonBridgman-Nearmap committed:
Added new unit tests to cover output of JsonCode field in database.
Serialisation of field to be updated in following commit.

JasonBridgman-Nearmap committed:
Added additional integration testing
More integration testing to verify state of objects extracted from the database.

JasonBridgman-Nearmap committed:
Hardened tests to check structure of data

JasonBridgman-Nearmap committed:
Removed Serializers from library.
- Moved serializers to test project. Will remove from test code in next commit.
- Added new models and automapper mappings

JasonBridgman-Nearmap committed:
Removed serializers from tests
- Updated tests to use automapper

JasonBridgman-Nearmap committed:
Make database interactions async
- Updated to use Async methods for database access
- Fixed some code warnings

JasonBridgman-Nearmap committed:
Update tests to make database interactions async

JasonBridgman-Nearmap committed:
Code refactor and clean
Also updated assembly info

drew-cooper commented on src/Core.Nhibernate/Core.Nhibernate.csproj:
```diff
@@ -31,37 +33,57 @@
+    <Reference Include="Newtonsoft.Json, Version=13.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed, processorArchitecture=MSIL">
+      <HintPath>..\packages\Newtonsoft.Json.13.0.1\lib\net45\Newtonsoft.Json.dll</HintPath>
+    </Reference>
+    <Reference Include="NHibernate, Version=5.3.0.0, Culture=neutral, PublicKeyToken=aa95f207798dfdb4, processorArchitecture=MSIL">
```
We're currently running NHibernate 5.1 in the Framework. I don't see a reason to go beyond that.

drew-cooper commented on src/Core.Nhibernate/Core.Nhibernate.csproj:
```diff
@@ -31,37 +33,57 @@
+    <Reference Include="Microsoft.Extensions.Logging.Abstractions, Version=6.0.0.1, Culture=neutral, PublicKeyToken=adb9793829ddae60, processorArchitecture=MSIL">
+      <HintPath>..\packages\Microsoft.Extensions.Logging.Abstractions.6.0.1\lib\net461\Microsoft.Extensions.Logging.Abstractions.dll</HintPath>
+    </Reference>
+    <Reference Include="Newtonsoft.Json, Version=13.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed, processorArchitecture=MSIL">
```
Is this update necessary? I think the changes in this project should be the minimum viable what we're trying to do. If the purpose is to allow `IdentityServer` to run on Framework v5.x, then we definitely should not be updating dependencies beyond the Framework level.

drew-cooper commented on src/Core.Nhibernate/Entities/ClientSecret.cs:
```diff
@@ -10,7 +10,7 @@ public class ClientSecret : BaseEntity<Guid>
         public virtual string Description { get; set; }
 
-        public virtual DateTimeOffset? Expiration { get; set; }
+        public virtual DateTime? Expiration { get; set; }
```
Why?

JasonBridgman-Nearmap replied:
Npgsql 6 reads the date out as DateTime in UTC, not as DateTimeOffset. You will get an invalid cast exception when using DateTimeOffset, and changing over to NodaTime (as recommended by Npgsql) requires more investigation and work.

drew-cooper replied:
Ok

drew-cooper commented on src/Core.Nhibernate.IntegrationTests/Core.Nhibernate.IntegrationTests.csproj:
```diff
@@ -30,70 +37,155 @@
+    <Reference Include="Moq, Version=4.17.0.0, Culture=neutral, PublicKeyToken=69f491c39445e920, processorArchitecture=MSIL">
+      <HintPath>..\packages\Moq.4.17.2\lib\net45\Moq.dll</HintPath>
+    </Reference>
+    <Reference Include="Newtonsoft.Json, Version=13.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed, processorArchitecture=MSIL">
```
I'd rather not go beyond Framework versions of dependencies.

drew-cooper commented on src/Core.Nhibernate.IntegrationTests/packages.config:
```diff
@@ -1,20 +1,48 @@
+  <package id="Fare" version="2.2.0" targetFramework="net472" />
+  <package id="FluentAssertions" version="6.7.0" targetFramework="net472" />
+  <package id="FluentNHibernate" version="3.1.0" targetFramework="net472" />
+  <package id="IdentityServer3" version="2.6.3" targetFramework="net472" />
+  <package id="Iesi.Collections" version="4.0.4" targetFramework="net472" />
+  <package id="Microsoft.Bcl.AsyncInterfaces" version="6.0.0" targetFramework="net472" />
+  <package id="Microsoft.Bcl.HashCode" version="1.1.1" targetFramework="net472" />
+  <package id="Moq" version="4.17.2" targetFramework="net472" />
+  <package id="Newtonsoft.Json" version="13.0.1" targetFramework="net472" />
+  <package id="NHibernate" version="5.3.11" targetFramework="net472" />
+  <package id="Npgsql" version="6.0.3" targetFramework="net472" />
+  <package id="Owin" version="1.0" targetFramework="net472" />
+  <package id="Remotion.Linq" version="2.2.0" targetFramework="net472" />
+  <package id="Remotion.Linq.EagerFetching" version="2.2.0" targetFramework="net472" />
+  <package id="System.Buffers" version="4.5.1" targetFramework="net472" />
+  <package id="System.Collections.Immutable" version="6.0.0" targetFramework="net472" />
+  <package id="System.Diagnostics.DiagnosticSource" version="6.0.0" targetFramework="net472" />
+  <package id="System.IO" version="4.3.0" targetFramework="net472" />
+  <package id="System.Memory" version="4.5.4" targetFramework="net472" />
+  <package id="System.Net.Http" version="4.3.4" targetFramework="net472" />
+  <package id="System.Numerics.Vectors" version="4.5.0" targetFramework="net472" />
+  <package id="System.Runtime" version="4.3.0" targetFramework="net472" />
+  <package id="System.Runtime.CompilerServices.Unsafe" version="6.0.0" targetFramework="net472" />
+  <package id="System.Security.Cryptography.Algorithms" version="4.3.0" targetFramework="net472" />
+  <package id="System.Security.Cryptography.Encoding" version="4.3.0" targetFramework="net472" />
+  <package id="System.Security.Cryptography.Primitives" version="4.3.0" targetFramework="net472" />
+  <package id="System.Security.Cryptography.X509Certificates" version="4.3.0" targetFramework="net472" />
+  <package id="System.Text.Encodings.Web" version="6.0.0" targetFramework="net472" />
+  <package id="System.Text.Json" version="6.0.2" targetFramework="net472" />
+  <package id="System.Threading.Channels" version="6.0.0" targetFramework="net472" />
+  <package id="System.Threading.Tasks.Extensions" version="4.5.4" targetFramework="net472" />
+  <package id="System.ValueTuple" version="4.5.0" targetFramework="net472" />
+  <package id="xunit" version="2.4.1" targetFramework="net472" />
+  <package id="xunit.abstractions" version="2.0.3" targetFramework="net472" />
+  <package id="xunit.analyzers" version="0.10.0" targetFramework="net472" />
+  <package id="xunit.assert" version="2.4.1" targetFramework="net472" />
+  <package id="xunit.core" version="2.4.1" targetFramework="net472" />
+  <package id="xunit.extensibility.core" version="2.4.1" targetFramework="net472" />
+  <package id="xunit.extensibility.execution" version="2.4.1" targetFramework="net472" />
+  <package id="xunit.runner.console" version="2.4.1" targetFramework="net472" developmentDependency="true" />
```
Is all of this necessary?

drew-cooper commented on src/Core.Nhibernate/Stores/NhibernateStore.cs:
```diff
@@ -24,71 +25,94 @@
                     try
                     {
-                        actionToExecute.Invoke(_nhSession);
-                        tx.Commit();
+                        await actionToExecute(_nhSession);
+                        await tx.CommitAsync();
                     }
                     catch (Exception)
                     {
-                        tx.Rollback();
+                        await tx.RollbackAsync();
                         throw;
                     }
                 }
 
             }
         }
 
-        protected T ExecuteInTransaction<T>(Func<ISession, T> actionToExecute, IsolationLevel isolationLevel = IsolationLevel.ReadCommitted)
+        protected async Task<T> ExecuteInTransactionAsync<T>(Func<ISession, Task<T>> actionToExecute, IsolationLevel isolationLevel = IsolationLevel.ReadCommitted)
```
When making these kinds of changes in a library you're better off adding the async version and marking the sync version `Obsolete`. In our case that would allow a 2-step update to Nearmap.IdentityServer to minimise risk.

JasonBridgman-Nearmap replied:
This method isn't directly used by our IdentityServer, for our use-case it's internal to this library only. I'll have a look at putting the non-asyncs back in though as an alternative for occasions where async is not desired for usage.

drew-cooper replied:
Sorry, I see what's happening now. The IdentityServer3 interfaces we're implementing here are aysnc and we're just pushing the async all the way down. Yeah, no need for the sync methods.

drew-cooper commented on src/Core.Nhibernate/Models/AuthorizationCode.cs:
```diff
@@ -0,0 +1,82 @@
+using System.Collections.Generic;
+using System.Linq;
+
+namespace IdentityServer3.Contrib.Nhibernate.Models
```
Where do these models come from? Why are they necessary?

JasonBridgman-Nearmap replied:
These have been created to allow the process previously being done by custom Json serialisers to be replaced with an Automapper object-object equivalent. The serialisers were writing directly to the database and performing object structure conversions in the process. This enables separation of object mapping, serialisation and IO operations properly.

drew-cooper reviewed (changes requested):
There's a lot going on here that is nice, but that I'm not sure is necessary. eg. async/await changes, injected AutoMapper, etc.
As a 3rd-party dependency I'd like to see a minimal viable update to allow building `IdentityServer` on FW v5.x (that's the main goal, right?), with as few API changes as possible. That would be an update, driven by a desire to support a newer version of NHibernate, that could be shared back to the community with minimal breaking changes.
Once we have that version then further changes to streamline things can be made, if necessary.

JasonBridgman-Nearmap committed:
COMM-455 Update build scripts and projects

JasonBridgman-Nearmap committed:
COMM-455 Update project files
- Streamline project files
- Remove unneeded nugets

JasonBridgman-Nearmap committed:
COMM-455 Cleanup of Core.Nhibernate project file
- Minor update to build script, don't include .pdb file in release nuget

JasonBridgman-Nearmap committed:
COMM-455 Remove injection of Automapper from Stores.

JasonBridgman-Nearmap committed:
COMM-455 Fix for datetime column type mapping

JasonBridgman-Nearmap committed:
COMM-455 Remove Automapper dependency from WebHost sample

JasonBridgman-Nearmap committed:
COMM-455 Reduced direct dependencies
- Removed direct references to NHibernate
- Changed FluentNHibernate to v2.1.2
- Update WebHost project file

Created packages.config at solution root for build scripts.

JasonBridgman-Nearmap committed:
COMM-455 Restrict class visibility
Refactoring to restrict visibility of classes
Removed unncessary IBaseEntity interface

JasonBridgman-Nearmap committed:
COMM-455 Added back ExecuteInTransaction methods
Returned non async methods to NHibernateStore and marked as obsolete for backward compatibility.

JasonBridgman-Nearmap committed:
COMM-455 Add driver tests
- Added new test projects to target Npgsql4 and Npgsql6 database drivers.
- Added configuration system to enable custom mappings if required by the database drivers.

JasonBridgman-Nearmap committed:
COMM-455 Fixing tests
- Accidentally broke test suites when test projects were split for v4 and v6 Npgsql packages.

JasonBridgman-Nearmap committed:
COMM-455 Added Extensions and Repository
- Added Autofac extension methods
- Added DataSourceRespository

JasonBridgman-Nearmap committed:
COMM-455 Steamline integration
Reduced complexity of integration by removing unncessary layers and adding extension methods.

JasonBridgman-Nearmap committed:
COMM-455 Update Extension
- Update RegisterNhibernateStores to require IDbProfileConfig parameter

drew-cooper commented on src/Core.Nhibernate.IntegrationTests/ObjectCreator.cs:
```diff
@@ -85,27 +92,30 @@ public static RefreshToken GetRefreshToken(string subjectId = null, string clien
+                .Without(t => t.Client)
+                .With(t => t.Claims, new List<Claim>()
+                {
+                    new Claim(Constants.ClaimTypes.Subject,subjectId ?? Guid.NewGuid().ToString())
```
Add space after `,`

drew-cooper commented on src/Core.Nhibernate.IntegrationTests/ObjectCreator.cs:
```diff
@@ -125,48 +135,29 @@ public static Client GetClient(string clientId = null)
-
-            return claims;
-        }
+            => Enumerable.Range(0, nClaimsToGet).Select(x => GetClaim());
```
Add a `.ToList()` here to avoid surprises if the result is enumerated more than once. If that happens you'll get a different set of claims every time.

drew-cooper commented on src/Core.Nhibernate.IntegrationTests/Stores/AuthorizationCodeStoreTests.cs:
```diff
@@ -24,283 +25,324 @@
+            }
+            jsonBuilder.Append("]");
+            jsonBuilder.Append("},");
+	        jsonBuilder.Append($"\"IsOpenId\":{code.IsOpenId.ToString().ToLowerInvariant()},");
```
Fix indent

JasonBridgman-Nearmap replied:
No longer relevant. Has now been replaced.

drew-cooper commented on src/Core.Nhibernate.IntegrationTests/Stores/AuthorizationCodeStoreTests.cs:
```diff
@@ -24,283 +25,324 @@
+                TokenType = TokenType.AuthorizationCode
+            };
+
+        private string GetJsonCodeFromAuthorizationCode(AuthorizationCode code)
         {
+            var jsonBuilder = new StringBuilder();
+
+            jsonBuilder.Append("{");
+            jsonBuilder.Append($"\"CreationTime\":\"{code.CreationTime:yyyy-MM-ddTHH:mm:ss.FFFFFFFzzz}\",");
+            jsonBuilder.Append("\"Client\":{");
+            jsonBuilder.Append($"\"ClientId\":\"{code.ClientId}\"");
+            jsonBuilder.Append("},");
+            jsonBuilder.Append("\"Subject\":{");
+            jsonBuilder.Append($"\"AuthenticationType\":\"{code.Subject.Identity.AuthenticationType}\",");
+            jsonBuilder.Append("\"Claims\":[");
+            foreach (var claim in code.Subject.Claims)
+            {
+                jsonBuilder.Append("{");
+                jsonBuilder.Append("\"Type\":\"sub\",");
+                jsonBuilder.Append($"\"Value\":\"{claim.Value}\"");
+                jsonBuilder.Append("}");
+            }
+            jsonBuilder.Append("]");
+            jsonBuilder.Append("},");
+	        jsonBuilder.Append($"\"IsOpenId\":{code.IsOpenId.ToString().ToLowerInvariant()},");
+            jsonBuilder.Append("\"RequestedScopes\":[");
```
This would be much cleaner using `Newtonsoft.Json.Linq`. Eg:
```c#
var json = new JObject(
    new JProperty("CreationTime", code.CreationTime),
    new JProperty("Client", new JObject(new JProperty("ClientId", code.ClientId)))
    new JProperty("Subject", new JObject(
        new JProperty("AuthenticationType", code.Subject.Identity.AuthenticationType),
        new JProperty("Claims", code.Subject.Claims.Select(c => 
            new JObject(new JProperty("Type", "sub"), new JProperty("Value", c.Value))))),
    new JProperty("IsOpenId", code.IsOpenId),
    new JProperty("RequestedScopes", code.RequestedScopes.Select(s => 
        new JObject(new JProperty("Name", s.Name)))),
    new JProperty("RedirectUri", code.RedirectUri),
    ...
    ...);

return json.ToString(Formatting.None);
```

JasonBridgman-Nearmap replied:
Now using dynamic object, which is even cleaner.

```
var obj = new
{
    CreationTime = code.CreationTime.ToString("yyyy-MM-ddTHH:mm:ss.FFFFFFFzzz"),
    Client = new { code.ClientId },
    Subject = new
    {
        code.Subject.Identity.AuthenticationType,
        Claims = code.Subject.Claims.Select(x => new { x.Type, x.Value }),
    },
    code.IsOpenId,
    RequestedScopes = code.RequestedScopes.Select(x => new { x.Name }),
    code.RedirectUri,
    code.Nonce,
    code.WasConsentShown,
    code.SessionId,
    code.CodeChallenge,
    code.CodeChallengeMethod,
    code.SubjectId,
    code.ClientId,
    Scopes = code.Scopes.Select(x => x)
};
```

JasonBridgman-Nearmap replied:
Hadn't meant to leave that quite that messy originally, thanks for pointing out.

drew-cooper commented on src/Core.Nhibernate.IntegrationTests/Stores/BaseStoreTests.cs:
```diff
@@ -27,123 +28,159 @@
+            ClientStore = new ClientStore(Session, dbProfile);
+        }
+
+        protected void RemoveTrailingComma(StringBuilder jsonBuilder)
```
Not needed if you use Newtonsoft.Json.Linq to build the Json string.

JasonBridgman-Nearmap replied:
Now using dynamic object, which is even cleaner.

drew-cooper commented on src/Core.Nhibernate.IntegrationTests/Stores/BaseStoreTests.cs:
```diff
@@ -27,123 +28,159 @@
+        {
+            if (jsonBuilder[jsonBuilder.Length - 1] == ',')
+            {
+                jsonBuilder.Remove(jsonBuilder.Length - 1, 1);
```
Simply decrementing the Length has the same effect.

JasonBridgman-Nearmap replied:
Have now gotten rid of this method as no longer required.

drew-cooper commented on src/Core.Nhibernate.IntegrationTests/Stores/RefreshTokenStoreTests.cs:
```diff
@@ -24,272 +25,331 @@
 
-            var testKey = Guid.NewGuid().ToString();
-            var testCode = ObjectCreator.GetRefreshToken();
+        private string GetJsonCodeFromRefreshToken(RefreshToken code)
```
See previous comment on using `Newtonsoft.Json.Linq`.

JasonBridgman-Nearmap replied:
Now using dynamic object, which is even cleaner.

drew-cooper commented on src/Core.Nhibernate.IntegrationTests/TestConstants.cs:
```diff
@@ -0,0 +1,9 @@
+{
+    internal static class TestConstants
+    {
+        public static readonly TimeSpan AllowableTimeVariance = new TimeSpan(0,0,1,0,0);
```
`TimeSpan.FromMinutes(1)`

drew-cooper commented on src/Core.Nhibernate/Core.Nhibernate.csproj:
```diff
@@ -1,142 +1,27 @@
   </ItemGroup>
-  <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
+  <ItemGroup>
+    <PackageReference Include="Autofac" Version="6.4.0" />
+    <PackageReference Include="AutoMapper" Version="10.0.0.0" />
+    <PackageReference Include="FluentNHibernate" Version="2.1.2" />
+    <PackageReference Include="IdentityServer3" Version="2.6.3" />
+    <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" Version="6.0.1" />
+    <PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
```
Please drop package versions to the version used in FW 5.1. 
I'd rather not have this package dictating the versions we're using.
```suggestion
    <PackageReference Include="Autofac" Version="6.2.0" />
    <PackageReference Include="AutoMapper" Version="10.1.1" />
    <PackageReference Include="FluentNHibernate" Version="2.1.2" />
    <PackageReference Include="IdentityServer3" Version="2.6.3" />
    <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" Version="6.0.1" />
    <PackageReference Include="Newtonsoft.Json" Version="12.0.3" />
```

JasonBridgman-Nearmap replied:
I'd like to update Autofac to 6.4.0 and Newtonsoft.Json to 13.0.1 (Newtonsoft particularly due to know issue prior to 13.0.1). AutoMapper version here isn't relevant as it's not included as a dependency for the library, it's merged into the DLL.

drew-cooper commented on src/Core.Nhibernate/Entities/BaseEntity.cs:
```diff
@@ -1,11 +1,8 @@
+    internal abstract class BaseEntity { }
 
-    public abstract class BaseEntity<TKey> : BaseEntity, IBaseEntity<TKey>
+    public abstract class BaseEntity<TKey>
```
Is there a reason for dropping the inheritance of `BaseEntity`?

JasonBridgman-Nearmap replied:
It's not necessary. The BaseEntity class exists as a Class with no generic <T> associated so that it can be used as an anchor by AutoMapper config setup. Other than that it isn't used.

drew-cooper commented on src/Core.Nhibernate/Entities/IBaseEntity.cs:
```diff
@@ -1,11 +0,0 @@
-    {
-    }
-
-    public interface IBaseEntity<TKey> : IBaseEntity
```
Is there a reason for dropping this?

JasonBridgman-Nearmap replied:
Ah, yes, it was only being used to allow AutoMapper to reference the assembly, so I dropped it and used the `BaseEntity` class to do the same thing.

drew-cooper commented on src/Core.Nhibernate/Extensions/AutofacExtensions.cs:
```diff
@@ -0,0 +1,154 @@
+        /// </summary>
+        /// <param name="builder">The <see cref="Autofac.ContainerBuilder"/></param>
+        /// <returns>The <see cref="Autofac.ContainerBuilder"/> instance for method chaining</returns>
+        public static ContainerBuilder UseNpgsql4ProviderConfig(this ContainerBuilder builder)
```
For Autofac extensions I'd rather `Register` be the verb instead of `Use`, for consistency with the rest of the API.

drew-cooper commented on src/Core.Nhibernate/NhibernateConfig/AutomappingConfiguration.cs:
```diff
@@ -65,8 +65,8 @@ public AutomappingConfiguration(bool registerOperationalServices, bool registerC
-                         && (_registerOperationalServices && _operationalServicesEntities.Contains(type)
-                             || _registerConfigurationServices && _configurationServicesEntities.Contains(type));
+                && (_registerOperationalServices && _operationalServicesEntities.Contains(type)
+                || _registerConfigurationServices && _configurationServicesEntities.Contains(type));
```
Indent this to show it's inside parens.

drew-cooper commented on src/Core.Nhibernate/Profiles/EntitiesProfile.cs:
```diff
@@ -0,0 +1,157 @@
+
+        private IEnumerable<Claim> GetClaims(IEnumerable<ContribModels.ClaimLite> claims)
+        {
+            return claims.Select(source => new Claim(source.Type, source.Value));
```
Add `.ToList()` to protect from multiple enumerations.

drew-cooper commented on src/Core.Nhibernate/Profiles/EntitiesProfile.cs:
```diff
@@ -0,0 +1,157 @@
+
+        private ClaimsPrincipal GetClaimsPrincipal(ContribModels.ClaimsPrincipalLite source)
+        {
+            var claims = source.Claims.Select(x => new Claim(x.Type, x.Value));
```
Add `.ToList()`

drew-cooper commented on src/Core.Nhibernate/Profiles/EntitiesProfile.cs:
```diff
@@ -0,0 +1,157 @@
+                    opt => opt.MapFrom(src => src.AllowAccessToAllGrantTypes))
+                .ForMember(x => x.AllowedCustomGrantTypes,
+                    opt => opt.MapFrom(src => src.AllowedCustomGrantTypes.Select(x => x.GrantType)))
+                .ForMember(x => x.RedirectUris, opt => opt.MapFrom(src => src.RedirectUris.Select(x => x.Uri)))
```
Add `.ToList()`

drew-cooper commented on src/Core.Nhibernate/Profiles/EntitiesProfile.cs:
```diff
@@ -0,0 +1,157 @@
+                    opt => opt.MapFrom(src => src.AllowedCustomGrantTypes.Select(x => x.GrantType)))
+                .ForMember(x => x.RedirectUris, opt => opt.MapFrom(src => src.RedirectUris.Select(x => x.Uri)))
+                .ForMember(x => x.PostLogoutRedirectUris,
+                    opt => opt.MapFrom(src => src.PostLogoutRedirectUris.Select(x => x.Uri)))
```
Add `.ToList()`

drew-cooper commented on src/Core.Nhibernate/Profiles/EntitiesProfile.cs:
```diff
@@ -0,0 +1,157 @@
+                .ForMember(x => x.PostLogoutRedirectUris,
+                    opt => opt.MapFrom(src => src.PostLogoutRedirectUris.Select(x => x.Uri)))
+                .ForMember(x => x.IdentityProviderRestrictions,
+                    opt => opt.MapFrom(src => src.IdentityProviderRestrictions.Select(x => x.Provider)))
```
Add `.ToList()`

drew-cooper commented on src/Core.Nhibernate/Profiles/EntitiesProfile.cs:
```diff
@@ -0,0 +1,157 @@
+                    opt => opt.MapFrom(src => src.PostLogoutRedirectUris.Select(x => x.Uri)))
+                .ForMember(x => x.IdentityProviderRestrictions,
+                    opt => opt.MapFrom(src => src.IdentityProviderRestrictions.Select(x => x.Provider)))
+                .ForMember(x => x.AllowedScopes, opt => opt.MapFrom(src => src.AllowedScopes.Select(x => x.Scope)))
```
Add `.ToList()`

drew-cooper commented on src/Core.Nhibernate/Profiles/EntitiesProfile.cs:
```diff
@@ -0,0 +1,157 @@
+                    opt => opt.MapFrom(src => src.IdentityProviderRestrictions.Select(x => x.Provider)))
+                .ForMember(x => x.AllowedScopes, opt => opt.MapFrom(src => src.AllowedScopes.Select(x => x.Scope)))
+                .ForMember(x => x.AllowedCorsOrigins,
+                    opt => opt.MapFrom(src => src.AllowedCorsOrigins.Select(x => x.Origin)))
```
Add `.ToList()`

drew-cooper commented on src/Core.Nhibernate/Profiles/EntitiesProfile.cs:
```diff
@@ -0,0 +1,157 @@
+                .ForMember(x => x.AllowedCorsOrigins,
+                    opt => opt.MapFrom(src => src.AllowedCorsOrigins.Select(x => x.Origin)))
+                .ForMember(x => x.Claims,
+                    opt => opt.MapFrom(src => src.Claims.Select(x => new Claim(x.Type, x.Value))));
```
Add `.ToList()`

drew-cooper commented on src/Core.Nhibernate/Profiles/EntitiesProfile.cs:
```diff
@@ -0,0 +1,157 @@
+                    opt => opt.MapFrom(src => src.AllowedScopes.Select(x => new Entities.ClientScope { Scope = x })))
+                .ForMember(x => x.AllowedCorsOrigins,
+                    opt => opt.MapFrom(
+                        src => src.AllowedCorsOrigins.Select(x => new Entities.ClientCorsOrigin { Origin = x })))
```
Add `.ToList()`

drew-cooper commented on src/Core.Nhibernate/Profiles/EntitiesProfile.cs:
```diff
@@ -0,0 +1,157 @@
+                        src => src.IdentityProviderRestrictions.Select(
+                            x => new Entities.ClientIdPRestriction { Provider = x })))
+                .ForMember(x => x.AllowedScopes,
+                    opt => opt.MapFrom(src => src.AllowedScopes.Select(x => new Entities.ClientScope { Scope = x })))
```
Add `.ToList()`

drew-cooper replied:
You could define a simple map from `string` to `Entities.ClientScope`, and then this complex piece would not be needed. Same is probably true for most of the enumerable mapping rules.

drew-cooper commented on src/Core.Nhibernate/Profiles/EntitiesProfileNpgsql4.cs:
```diff
@@ -0,0 +1,22 @@
+
+            CreateMap<CoreModels.Secret, Entities.ClientSecret>(MemberList.Source)
+                .ForMember(x => x.Expiration, opts => opts.MapFrom(
+                    src => src.Expiration.HasValue ? (DateTime?)src.Expiration.Value.DateTime : null));
```
Just add the following mapping and you don'tneed the custom rules for these maps:
```c#
CreateMap<DateTimeOffset, DateTime>().ConstructUsing(dto => dto.DateTime);
```

drew-cooper commented on src/Core.Nhibernate/Profiles/EntitiesProfileNpgsql6.cs:
```diff
@@ -0,0 +1,22 @@
+    {
+        public EntitiesProfileNpgSql6() : base()
+        {
+            CreateMap<CoreModels.Secret, Entities.ScopeSecret>(MemberList.Source)
+                .ForMember(x => x.Expiration, opts => opts.MapFrom(
+                    src => src.Expiration.HasValue ? (DateTime?)src.Expiration.Value.DateTime.ToUniversalTime() : null));
+
+            CreateMap<CoreModels.Secret, Entities.ClientSecret>(MemberList.Source)
+                .ForMember(x => x.Expiration, opts => opts.MapFrom(
+                    src => src.Expiration.HasValue ? (DateTime?)src.Expiration.Value.DateTime.ToUniversalTime() : null));
```
Just add the following mapping and you don'tneed the custom rules for these maps:
```c#
CreateMap<DateTimeOffset, DateTime>().ConstructUsing(dto => dto.DateTime.ToUniversalTime());
```

drew-cooper replied:
Also, `.DateTime.ToUniversalTime()` will convert the time regardless of the original time zone. 
`.UtcDateTime` will give a UTC time, only converting the time if necessary. 
Which behaviour do you want here? The second is often correct.

drew-cooper replied:
It would be good to have tests of these maps (And the Npgsql4 ones) that assert the correct behaviour.

JasonBridgman-Nearmap replied:
Thanks for those tips. The Mapping behaviors are being tested, that's why there is 2 distinct test projects, one running against Npgsql 4, the other Npgsql 6.

drew-cooper commented on src/Core.Nhibernate/Stores/NhibernateStore.cs:
```diff
@@ -24,71 +25,114 @@
+        protected NhibernateStore(ISession session, IDbProfileConfig dbProfile)
         {
-            if (session == null) throw new ArgumentNullException(nameof(session));
+            _mapper = new MapperConfiguration(cfg =>
+            {
+                cfg.AddProfile(dbProfile?.GetProfile() ?? new EntitiesProfile());
+            })
+                .CreateMapper();
```
Building the configuration is expensive, and is intended to be done once. Creating a mapper from a config is cheap.
Can you use `builder.RegisterAutoMapper()` in the Autofac setup and inject the `IMapper` here?
Or register a complete configuraiton (or a provider thereof) and inject that here and then create the mapper.

JasonBridgman-Nearmap replied:
Have updated approach

drew-cooper commented on src/Core.Nhibernate/Stores/AuthorizationCodeStore.cs:
```diff
@@ -23,41 +24,121 @@
+                if (token == null)
+                {
+                    return null;
+                }
```
Inlining trival blocks like this is nice to save vertical space.

drew-cooper replied:
ie. `if (token == null) { return null; }`

drew-cooper commented on src/Core.Nhibernate/Stores/AuthorizationCodeStore.cs:
```diff
@@ -23,41 +24,121 @@
+            await SaveAsync(nhCode);
 
-            await TaskExtensions.CompletedTask;
+            await Task.CompletedTask;
```
No need for this.

JasonBridgman-Nearmap replied:
Thanks

drew-cooper commented on src/Core.Nhibernate/Stores/ClientStore.cs:
```diff
@@ -33,23 +33,30 @@ namespace IdentityServer3.Contrib.Nhibernate.Stores
-            return await Task.FromResult(client);
+        public object Save(Core.Models.Client obj)
+        {
+            return SaveAsync(_mapper.Map<Client>(obj)).Result;
```
If you have to run a task synchronously do `.GetAwaiter().GetResult();`. It does the right thing with exceptions.
But prefer blocking all the away down.
If you need the non-async Save method here then have a non-async Save method in the base which does non-async transaction and `session.Save()`;

drew-cooper commented on src/Core.Nhibernate/Stores/ScopeStore.cs:
```diff
@@ -28,56 +29,59 @@
-              });
+        public object Save(Core.Models.Scope obj)
+        {
+            return SaveAsync(_mapper.Map<Scope>(obj)).Result;
```
See other comment on similar code.

drew-cooper commented on src/Core.Nhibernate/Stores/ConsentStore.cs:
```diff
@@ -35,126 +35,94 @@ namespace IdentityServer3.Contrib.Nhibernate.Stores
+                {
+                    Subject = item.Subject,
+                    ClientId = item.ClientId,
+                    Scopes = ParseScopes(item.Scopes)
```
Does `items.Scope` need to be inside the transaction?

JasonBridgman-Nearmap replied:
Updated to make a request to the Scopes reference whilst in the transaction.

drew-cooper commented on src/Core.Nhibernate/Stores/ConsentStore.cs:
```diff
@@ -35,126 +35,94 @@ namespace IdentityServer3.Contrib.Nhibernate.Stores
 
-        public async Task UpdateAsync(IdentityServer3.Core.Models.Consent consent)
+        public async Task UpdateAsync(Core.Models.Consent consent)
+            => await ExecuteInTransactionAsync(session => UpdateInnerAsync(session, consent));
```
The lambda should be async/await. And I'm not sure the inner method helps here.

JasonBridgman-Nearmap replied:
have added async/await, though only needed for call stack traces.

The Inner method makes the code easier to read, otherwise it would need to be a lambda in the `ExecuteInTransactionAsync` method, and that's just ugly for a method the size of the one being used.

drew-cooper commented on src/Core.Nhibernate/Stores/ConsentStore.cs:
```diff
@@ -35,126 +35,94 @@ namespace IdentityServer3.Contrib.Nhibernate.Stores
-
-                    session.Save(item);
+                    return;
                 }
```
inline the trivial block

drew-cooper commented on src/Core.Nhibernate/Stores/ConsentStore.cs:
```diff
@@ -35,126 +35,94 @@ namespace IdentityServer3.Contrib.Nhibernate.Stores
-        private IEnumerable<string> ParseScopes(string scopes)
-        {
-            return string.IsNullOrWhiteSpace(scopes) ? Enumerable.Empty<string>() : scopes.Split(',');
+                
```
Remove extra blank line

drew-cooper commented on src/Core.Nhibernate/Stores/DataSourceRepository.cs:
```diff
@@ -0,0 +1,55 @@
+
+namespace IdentityServer3.Contrib.Nhibernate.Stores
+{
+    public class DataSourceRepository : IDataSourceRepository
```
Why to we need this as distinct from the main container?

drew-cooper commented on src/Core.Nhibernate/Stores/NhibernateStore.cs:
```diff
@@ -24,71 +25,114 @@
+            => ExecuteInTransactionAsync(
+                session => { actionToExecute(session); return Task.CompletedTask; },
+                isolationLevel)
+            .GetAwaiter().GetResult();
```
Don't do this. Call down to NHibernate's synchronous methods.

JasonBridgman-Nearmap replied:
Have removed this now.

drew-cooper commented on src/Core.Nhibernate/Stores/RefreshTokenStore.cs:
```diff
@@ -30,45 +31,116 @@
 
-        public override async Task StoreAsync(string key, RefreshToken value)
+        public override async Task<RefreshToken> GetAsync(string key)
+            => await ExecuteInTransactionAsync(session => GetInnerAsync(session, key));
+
+        private async Task<RefreshToken> GetInnerAsync(ISession session, string key)
```
Not sure the inner method here buys us anything.

JasonBridgman-Nearmap replied:
Used for readability.

drew-cooper commented on src/Core.Nhibernate/Stores/TokenHandleStore.cs:
```diff
@@ -23,40 +24,112 @@
+            if (token == null)
+            {
+                return null;
+            }
```
inline

drew-cooper commented on src/IdentityServer3.Contrib.Nhibernate.nuspec:
```diff
@@ -1,24 +1,27 @@
+	<tags>IdentityServer OpenID Connect OpenIDConnect OAuth2 OWIN ASP.NET Katana WebApi SSO Federation Claims Identity Nhibernate</tags>
+    <dependencies>
+	  <group targetFramework=".NETFramework4.7.2">
+		<dependency id="Autofac" version="6.4.0" />
+		<dependency id="FluentNHibernate" version="2.1.2" />
+		<dependency id="IdentityServer3" version="2.6.3" />
+		<dependency id="Newtonsoft.Json" version="13.0.1" />
+		<dependency id="Microsoft.Extensions.Logging.Abstractions" version="6.0.1" />
```
Fix dependency versions per comments on csproj.

JasonBridgman-Nearmap replied:
Considering COMM-464 (issues with Newtonsoft.Json prior to 13.0.1), are we intending to bump Framework to v13.0.1 of this package? The only other dependency package not currently running the same version is Autofac, which we currently have on 6.2.0 in Framework. I would prefer to bump these two up in Framework if there is no strong objection.

The version of Automapper being used currently isn't relevant as its merged into the package.

drew-cooper commented on src/Core.Nhibernate/Extensions/IdentityServerServiceFactoryExtensions.cs:
```diff
@@ -1,91 +1,65 @@
             {
-                factory.Register(
-                    new Registration<ISessionFactory>(serviceOptions.NhibernateSessionFactory));
+                factory.Register(new Registration<IDbProfileConfig>(dbProfileConfig));
```
Should probably build the configuration here and and register the IMapper instead. Unless there's a more appropriate location.

JasonBridgman-Nearmap replied:
Have updated the approach.

drew-cooper reviewed (changes requested):
A few things need to be fixed up:
* AutoMapper config should be built as early as possible and then the `IMapper` instance injected where it's needed.
* I understand now that the change to async is because the interfaces being implemented are async. No need then to maintain the sync methods. Sorry about that.
* Dependency versions should not be greater than those in FW 5.1.
* Other stuff per comments.

drew-cooper commented on src/Core.Nhibernate/Extensions/IdentityServerServiceFactoryExtensions.cs:
```diff
@@ -1,91 +1,65 @@
+            ISessionFactory nHibernateSessionFactory,
+            IDbProfileConfig dbProfileConfig, 
+            bool registerOperationalServices = false, 
+            bool registerConfigurationServices = false)
```
I would prefer to keep the parameters object. It makes the API easier to use because it's harder to confuse the 2 bool parameters.
The AutoMapper piece could be added to it.

drew-cooper commented on src/Core.Nhibernate/Profiles/IDbProfileConfig.cs:
```diff
@@ -0,0 +1,9 @@
+
+namespace IdentityServer3.Core.Models
+{
+    public interface IDbProfileConfig
+    {
+        Profile GetProfile();
```
I don't like the naming of this. It's extremely non-obvious that it has anything to do with AutoMapper.

drew-cooper replied:
Maybe `IMappingProfileProvider`?

JasonBridgman-Nearmap replied:
Have now removed as no longer required with revised approach.

drew-cooper commented on src/Core.Nhibernate/Profiles/Npgsql4ProviderConfig.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    public class Npgsql4ProviderConfig : IDbProfileConfig
+    {
+        public Profile GetProfile()
+        {
+            return new EntitiesProfileNpgSql4();
+        }
```
Use expression body.

JasonBridgman-Nearmap committed:
COMM-455 Apply feedback from peer review
- Re-remove sync transaction wrappers
- Cleanup data structure verifiers
- Simplify AutoMapper mappings
- Remove DbProfile Config classes and Interface, no longer required, design change.

JasonBridgman-Nearmap committed:
COMM-455 Updated nupsec and build files to add AutoMapper as dependency.

JasonBridgman-Nearmap committed:
COMM-455 Updated WebHost project
Wasn't previously starting, now fixed.

JasonBridgman-Nearmap committed:
COMM-455 Removed DataSourceRepository
No extensive current benefit to this one, so removing for now.

JasonBridgman-Nearmap committed:
COMM-455 Removed tests missed from previous commit

JasonBridgman-Nearmap committed:
COMM-455 Updated MappingHelper
- Simplified code
- Added optional conventions mapping to GetNhibernateServicesMappings method

drew-cooper commented on src/Core.Nhibernate/NhibernateConfig/MappingHelper.cs:
```diff
@@ -51,20 +60,11 @@ public static AutoPersistenceModel GetNhibernateServicesMappings(bool registerOp
-        }
+        public static IMapper CreateMapper(params Profile[] profileConfigs)
+            => new MapperConfiguration(cfg => cfg.AddProfiles(profileConfigs))
+                .CreateMapper();
```
I don't like this here. It's extending the existing helper into a completely different mapping context.
Use the `container.RegisterAutomapper(...)` method in `Program.cs` and maybe a test-specific helper for the tests.

drew-cooper commented on src/Core.Nhibernate/Profiles/EntitiesProfileNpgsql4.cs:
```diff
@@ -10,13 +7,7 @@ public class EntitiesProfileNpgSql4 : EntitiesProfile
     {
         public EntitiesProfileNpgSql4() : base()
```
` : base()` is redundant

drew-cooper commented on src/Core.Nhibernate/Profiles/EntitiesProfileNpgsql6.cs:
```diff
@@ -10,13 +7,7 @@ public class EntitiesProfileNpgSql6 : EntitiesProfile
     {
         public EntitiesProfileNpgSql6() : base()
```
`: base()` is redundant

drew-cooper commented on src/Core.Nhibernate/Extensions/AutofacExtensions.cs:
```diff
@@ -1,30 +1,21 @@
+        /// <param name="profileConfigs">AutoMapper profiles to apply to the mapper</param>
+        /// <returns>A <see cref="ContainerBuilder"/></returns>
+        public static ContainerBuilder RegisterAutoMapper(this ContainerBuilder builder, params Profile[] profileConfigs)
         {
-            builder.RegisterType<Npgsql6ProviderConfig>().As<IDbProfileConfig>();
+            builder.RegisterInstance(MappingHelper.CreateMapper(profileConfigs));
             return builder;
         }
```
There is already a `RegisterAutoMapper` method in `AutoMapper.Contrib.Autofac.DependencyInjection`

JasonBridgman-Nearmap replied:
Yes, but not one that does this.

drew-cooper commented on src/Core.Nhibernate/Stores/ConsentStore.cs:
```diff
@@ -35,17 +36,22 @@ namespace IdentityServer3.Contrib.Nhibernate.Stores
+                var consent = await session.Query<Consent>()
+                    .SingleOrDefaultAsync(c => c.Subject == subject && c.ClientId == client);
+
+                _ = consent?.Scopes; // Force access of the Scopes link within the transaction.
```
Should eager loading of this be configured in the mapping?
Or use `if (consent is not null) { NHubernateUtil.Initialize(consent.Scopes); }`

JasonBridgman-Nearmap replied:
Nope, I should have paid attention here, `consent.Scopes` is a string, not a collection.

drew-cooper commented on src/WebHost/Config/Factory.cs:
```diff
@@ -74,38 +81,38 @@ private static void BuildSchema(Configuration cfg)
+                    await clientStore.SaveAsync(client);
                 }
 
                 tx.Commit();
```
`await tx.CommitAsync()`

JasonBridgman-Nearmap replied:
Thanks

drew-cooper commented on src/WebHost/Config/Factory.cs:
```diff
@@ -74,38 +81,38 @@ private static void BuildSchema(Configuration cfg)
+                    await scopeStore.SaveAsync(scope);
                 }
 
                 tx.Commit();
```
`await tx.CommitAsync();`

drew-cooper commented on src/WebHost/Startup.cs:
```diff
@@ -21,17 +22,17 @@ public void Configuration(IAppBuilder appBuilder)
                     SiteName = "IdentityServer3 (Nhibernate)",
                     SigningCertificate = Certificate.Get(),
-                    Factory = Factory.Configure(logger)
+                    Factory = await Factory.ConfigureAsync(logger)//.GetAwaiter().GetResult()
```
Remove commented code

JasonBridgman-Nearmap replied:
Doh, thanks, I usually pick those things up during staging reviews.

drew-cooper commented on src/WebHost/Web.config:
```diff
@@ -42,13 +39,29 @@
   </appSettings>
   <runtime>
     <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
+			<dependentAssembly>
+				<assemblyIdentity name="System.Threading.Tasks.Extensions" publicKeyToken="CC7B13FFCD2DDD51" culture="neutral" />
+				<bindingRedirect oldVersion="0.0.0.0-4.2.0.1" newVersion="4.2.0.1" />
+			</dependentAssembly>
+			<dependentAssembly>
+				<assemblyIdentity name="System.Memory" publicKeyToken="CC7B13FFCD2DDD51" culture="neutral" />
+				<bindingRedirect oldVersion="0.0.0.0-4.0.1.1" newVersion="4.0.1.1" />
+			</dependentAssembly>
+			<dependentAssembly>
+				<assemblyIdentity name="System.Diagnostics.DiagnosticSource" publicKeyToken="CC7B13FFCD2DDD51" culture="neutral" />
+				<bindingRedirect oldVersion="0.0.0.0-6.0.0.0" newVersion="6.0.0.0" />
+			</dependentAssembly>
+			<dependentAssembly>
+				<assemblyIdentity name="Microsoft.Bcl.AsyncInterfaces" publicKeyToken="CC7B13FFCD2DDD51" culture="neutral" />
+				<bindingRedirect oldVersion="0.0.0.0-6.0.0.0" newVersion="6.0.0.0" />
+			</dependentAssembly>
```
nit: fix indenting

JasonBridgman-Nearmap replied:
Thanks, it ended up with tabs in there somehow.

drew-cooper reviewed (changes requested):
Getting there. Just a few more tweaks.

JasonBridgman-Nearmap committed:
COMM-455 Apply Peer review
- Update with peer review feedback
- Move db config items into dedicated Config classes

richarddbarnett commented on src/Core.Nhibernate.IntegrationTests/Core.Nhibernate.IntegrationTests.csproj:
```diff
@@ -30,49 +33,93 @@
+      <HintPath>..\packages\NHibernate.5.3.11\lib\net461\NHibernate.dll</HintPath>
+    </Reference>
+    <Reference Include="Npgsql, Version=6.0.3.0, Culture=neutral, PublicKeyToken=5d8b90d52f46fda7, processorArchitecture=MSIL">
+      <HintPath>..\packages\Npgsql.6.0.3\lib\netstandard2.0\Npgsql.dll</HintPath>
```
is it safe to go to Npgsql 6.x when we're still using 4.x in Framework 5.1 & 6.0? (& IdentityServer is on 3.2.7)

JasonBridgman-Nearmap replied:
The library handles both 4.x (as per current version of framework) and 6.x (as the latest available). There is a dedicated test project for each driver version.

richarddbarnett commented on src/Core.Nhibernate.IntegrationTests/Stores/AuthorizationCodeStoreTests.cs:
```diff
@@ -1,6 +1,7 @@
 ﻿/*MIT License
 *
 *Copyright (c) 2016 Ricardo Santos
+*Copyright (c) 2022 Jason F. Bridgman
```
Copyright should be Nearmap, or unchanged.

richarddbarnett commented on src/Core.Nhibernate.IntegrationTests/Stores/AuthorizationCodeStoreTests.cs:
```diff
@@ -145,8 +138,9 @@ public async Task RemoveAsync()
                 var token = session.Query<Token>()
-                    .SingleOrDefault(t => t.TokenType == TokenType.AuthorizationCode &&
-                                          t.Key == testKey);
+                    .SingleOrDefault(t => 
```
no need for this change - or can put the whole `.SingleOrDefault()` on 1 line

richarddbarnett commented on src/Core.Nhibernate.IntegrationTests/Stores/BaseStoreTests.cs:
```diff
@@ -62,6 +65,12 @@ public abstract class BaseStoreTests
+            {
+                cfg.AddProfile<EntitiesProfile>();
+            })
+                .CreateMapper();
```
to single-line, or fix indent

richarddbarnett commented on src/Core.Nhibernate/Extensions/ClientExtensions.cs:
```diff
@@ -0,0 +1,49 @@
+            if (s.AllowedCorsOrigins == null)
+            {
+                s.AllowedCorsOrigins = new List<string>();
+            }
```
not keen on mutating the `Client`.
can this logic be performed by automapper?
if it must be kept, is it better on a single line?
```csharp
s.AllowedCorsOrigins = s.AllowedCorsOrigins ?? new List<string>();
```

richarddbarnett replied:
Can't https://docs.automapper.org/en/stable/Lists-and-arrays.html#handling-null-collections be used instead?

JasonBridgman-Nearmap replied:
File no longer exists

richarddbarnett commented on src/WebHost/Startup.cs:
```diff
@@ -17,16 +17,21 @@ public void Configuration(IAppBuilder appBuilder)
             Log.Logger = new LoggerConfiguration()
                .MinimumLevel.Debug()
                .WriteTo.Trace()
-               .WriteTo.RollingFile(@"C:\Users\RicardoSantos\Repos\Projects\IdentityServer3.Contrib.Nhibernate\src\WebHost\log\Log-{Date}.log")
```
👍

richarddbarnett commented on src/Core.Nhibernate.IntegrationTests/Stores/BaseStoreTests.cs:
```diff
@@ -33,22 +33,17 @@
 using NHibernate.Tool.hbm2ddl;
-using NHibernate.Type;
 using Configuration = NHibernate.Cfg.Configuration;
+using Id3Postgres = IdentityServer3.Contrib.Nhibernate.Postgres;
```
Prefer `IdSrv3` as an abbreviation for `IdentityServer3`

JasonBridgman-Nearmap replied:
This was removed

richarddbarnett commented on src/Core.Nhibernate.IntegrationTests/packages.config:
```diff
@@ -5,7 +5,7 @@
   <package id="AutoMapper" version="10.1.1" targetFramework="net472" />
   <package id="Castle.Core" version="4.4.1" targetFramework="net472" />
-  <package id="Fare" version="2.1.1" targetFramework="net472" />
+  <package id="Fare" version="2.2.0" targetFramework="net472" />
```
what uses this - AutoMapper?

JasonBridgman-Nearmap replied:
I'm not sure where this compare has come from, this file has been deleted...

richarddbarnett commented on src/Core.Nhibernate/packages.config:
```diff
@@ -5,9 +5,14 @@
+  <package id="System.Buffers" version="4.5.1" targetFramework="net472" />
+  <package id="System.Memory" version="4.5.4" targetFramework="net472" />
+  <package id="System.Numerics.Vectors" version="4.5.0" targetFramework="net472" />
+  <package id="System.Runtime.CompilerServices.Unsafe" version="4.5.3" targetFramework="net472" />
```
Are these necessary?

JasonBridgman-Nearmap replied:
? File has been deleted

richarddbarnett commented on src/Core.Nhibernate.IntegrationTests/Stores/ScopeStoreTests.cs:
```diff
@@ -83,27 +82,12 @@ public async Task FindScopesAsync()
-            //CleanUp
-            ExecuteInTransaction(session =>
-            {
-                session.Delete(testScope1Entity);
```
why is it safe to remove these entity deletes?

JasonBridgman-Nearmap replied:
? Not sure what your meaning here. The entity deletes are still there, the calls have been updated to be async.

richarddbarnett commented on src/Core.Nhibernate.IntegrationTests/Stores/AuthorizationCodeStoreTests.cs:
```diff
@@ -83,6 +137,34 @@ public async Task StoreAsync()
+        [Fact]
+        public async Task VerifyJsonCodeDataStructure()
+        {
+            // Setup
```
no need for setup/act/assert/cleanup comments.

JasonBridgman-Nearmap replied:
These have just been left in place. Can remove later.

richarddbarnett commented on src/Core.Nhibernate.IntegrationTests/Stores/AuthorizationCodeStoreTests.cs:
```diff
@@ -136,7 +136,10 @@ public async Task StoreAsync()
+                token.Should().NotBeNull();
+                token.TokenType.Should().Be(TokenType.AuthorizationCode);
+                token.Key.Should().Be(testKey);
+                token.ClientId.Should().Be(testCode.ClientId);
```
why not use FluentAssertions?

JasonBridgman-Nearmap replied:
Ah, this was potentially done in earlier passes, will update.

richarddbarnett commented on src/Core.Nhibernate.IntegrationTests/Stores/RefreshTokenStoreTests.cs:
```diff
@@ -254,8 +279,15 @@ public async Task GetAllAsync()
-            Assert.True(tokens.Count == 2);
-            Assert.True(tokens.All(t => t.SubjectId == subjectId1));
+            tokens.Should().HaveCount(2)
+                .And.AllBeOfType<RefreshToken>()
```
not sure you need these 2 assertions

richarddbarnett commented on src/Core.Nhibernate/Stores/AuthorizationCodeStore.cs:
```diff
@@ -42,14 +48,95 @@ public AuthorizationCodeStore(ISession session, IScopeStore scopeStore, IClientS
+                    if (code == null)
+                    {
+                        continue;
+                    }
```
if `_mapper.Map()` can't return `null` then this could be:
```csharp
if (token.Expiry < DateTime.UtcNow) { continue; }
var code = _mapper.Map<AuthorizationCode>(authCode);
```

JasonBridgman-Nearmap replied:
Mapper will return null when given a null. Have moved the date/time filter to the db call as it made more sense for it to filter there anyway.

richarddbarnett commented on src/Core.Nhibernate/Stores/AuthorizationCodeStore.cs:
```diff
@@ -42,14 +48,95 @@ public AuthorizationCodeStore(ISession session, IScopeStore scopeStore, IClientS
+                        continue;
+                    }
+
+                    code.Client = ClientStore.FindClientByIdAsync(authCode.ClientId).Result;
```
```csharp
code.Client = await ClientStore.FindClientByIdAsync(authCode.ClientId);
```
etc

JasonBridgman-Nearmap replied:
Already done

richarddbarnett commented on src/Core.Nhibernate/Stores/RefreshTokenStore.cs:
```diff
@@ -42,6 +48,80 @@ public RefreshTokenStore(ISession session, IScopeStore scopeStore, IClientStore
+                    return null;
+                }
+
+                var refToken = ConvertFromJson<RefToken>(token.JsonCode);
```
"RefToken" isn't a good name because of the ambiguity of "ref".
would  prefer something like `NHibRefreshToken`

richarddbarnett commented on src/Core.Nhibernate/Stores/AuthorizationCodeStore.cs:
```diff
@@ -42,14 +48,95 @@ public AuthorizationCodeStore(ISession session, IScopeStore scopeStore, IClientS
+                    new ClaimsIdentity(
+                        claims, authCode.Subject.AuthenticationType, Constants.ClaimTypes.Name, Constants.ClaimTypes.Role));
+
+                return code;
```
extract this to a method & call from `GetAllAsync`

richarddbarnett commented on src/Core.Nhibernate/Stores/RefreshTokenStore.cs:
```diff
@@ -42,6 +48,80 @@ public RefreshTokenStore(ISession session, IScopeStore scopeStore, IClientStore
+                    new ClaimsIdentity(
+                        claims, refToken.Subject.AuthenticationType, Constants.ClaimTypes.Name, Constants.ClaimTypes.Role));
+
+                return refreshToken;
```
extract to a method & call from `GetAllAsync`

richarddbarnett commented on src/Core.Nhibernate/Stores/TokenHandleStore.cs:
```diff
@@ -24,33 +24,111 @@
+
+                foreach(var token in tokens)
+                {
+                    var tModel = ConvertFromJson<NHibTokenModel>(token.JsonCode);
```
use a better name than `tModel`

richarddbarnett commented on src/Core.Nhibernate/Stores/TokenHandleStore.cs:
```diff
@@ -24,33 +24,111 @@
+                {
+                    var tModel = ConvertFromJson<NHibTokenModel>(token.JsonCode);
+
+                    if (tModel == null)
```
did you miss the equivalent of line 67?

JasonBridgman-Nearmap replied:
Judging by the other stores, I'm inclined to say yes, thanks for the pickup. Would still have worked, but would have slowed it down when many out of date tokens existed in the db.

richarddbarnett commented on src/Core.Nhibernate/Postgres/PostgreSQLConfiguration.cs:
```diff
@@ -3,6 +3,9 @@
 {
+    [System.Diagnostics.CodeAnalysis.SuppressMessage(
+        "Minor Code Smell", "S101:Types should be named in PascalCase", 
+        Justification = "Matching Heirarcy")]
```
typo "Hierarchy"

richarddbarnett commented on src/IdentityServer3.Contrib.Nhibernate.nuspec:
```diff
@@ -1,24 +1,28 @@
+    <id>IdentityServer3.Contrib.Nhibernate</id>
+    <version>0.0.0</version>
+    <title>IdentityServer3.Contrib.Nhibernate</title>
+    <authors>Ricardo Santos, Jason F. Bridgman and contributors</authors>
```
not sure if we need to check with Legal - but IMO instead of your name this should either be "Nearmap" or just left unchanged.

richarddbarnett commented on src/Core.Nhibernate.Tests/AutofacExtensionTests.cs:
```diff
@@ -0,0 +1,91 @@
+
+            container.Resolve<TInterface>().Should().BeOfType<TClass>().And.NotBeNull();
+
+            return new AndConstraint<IContainer>(container);
```
```csharp
return builder.Build().ShouldResolve<TInterface, TClass>();
```

richarddbarnett reviewed (changes requested):
Impressive work.

A few minor comments.

Why is `VersionInfo.cs` binary (commit 734a5d89bc04233b3a41d66e893970e70b920cb1)?

What does the WebHost project do?

I can't build this on a Mac:

`dotnet build` & `dotnet clean` both fail with
```
/Users/richard.barnett/nearmap/repos/forks/IdentityServer3.Contrib.Nhibernate/src/WebHost/WebHost.csproj(96,3): error MSB4019: The imported project "/usr/local/share/dotnet/x64/sdk/6.0.301/Microsoft/VisualStudio/v17.0/WebApplications/Microsoft.WebApplication.targets" was not found. Confirm that the expression in the Import declaration "/usr/local/share/dotnet/x64/sdk/6.0.301/Microsoft/VisualStudio/v17.0/WebApplications/Microsoft.WebApplication.targets" is correct, and that the file exists on disk.
```

In Rider the build failure is:
```
  WebHost.csproj(120, 5): This project references NuGet package(s) that are missing on this computer. Use NuGet Package Restore to download them.  For more information, see http://go.microsoft.com/fwlink/?LinkID=322105. The missing file is ../packages/Microsoft.CodeDom.Providers.DotNetCompilerPlatform.2.0.1/build/net46/Microsoft.CodeDom.Providers.DotNetCompilerPlatform.props.
```

JasonBridgman-Nearmap commented:
> Impressive work.
> 
> A few minor comments.
> 
> Why is `VersionInfo.cs` binary (commit [734a5d8](https://github.com/nearmap/IdentityServer3.Contrib.Nhibernate/commit/734a5d89bc04233b3a41d66e893970e70b920cb1))?
> 
> What does the WebHost project do?
> 
> I can't build this on a Mac:
> 
> `dotnet build` & `dotnet clean` both fail with
> 
> ```
> /Users/richard.barnett/nearmap/repos/forks/IdentityServer3.Contrib.Nhibernate/src/WebHost/WebHost.csproj(96,3): error MSB4019: The imported project "/usr/local/share/dotnet/x64/sdk/6.0.301/Microsoft/VisualStudio/v17.0/WebApplications/Microsoft.WebApplication.targets" was not found. Confirm that the expression in the Import declaration "/usr/local/share/dotnet/x64/sdk/6.0.301/Microsoft/VisualStudio/v17.0/WebApplications/Microsoft.WebApplication.targets" is correct, and that the file exists on disk.
> ```
> 
> In Rider the build failure is:
> 
> ```
>   WebHost.csproj(120, 5): This project references NuGet package(s) that are missing on this computer. Use NuGet Package Restore to download them.  For more information, see http://go.microsoft.com/fwlink/?LinkID=322105. The missing file is ../packages/Microsoft.CodeDom.Providers.DotNetCompilerPlatform.2.0.1/build/net46/Microsoft.CodeDom.Providers.DotNetCompilerPlatform.props.
> ```

The WebHost is effectively a sample project hosting IdentityServer with the Contrib library used. I've used powershell to build this on my local using the build.ps1 file. Not sure why the VersionInfo.cs file is binary... I haven't intentionally committed that yet, it may have slipped through at some point. Will fix. The framework version specified in the old project file used by the WebHost project I suspect isn't compatible with dotnet build, from what I've seen, it only supports the newer framework configs, so `<TargetFrameworkVersion>v4.7.2</TargetFrameworkVersion>` isn't supported, but `<TargetFramework>net472</TargetFramework>` is. Unfortunately, I found this project file was a bit more temperamental and resistant to being updated, so I've left it with minimal changes. The build.ps1 script is capable of building it though, but you may need to install the visual studio build tools. Unfortunately, not having a mac I can only provide very limited suggestions on this one.

JasonBridgman-Nearmap committed:
COMM-455 Apply peer review feedback

JasonBridgman-Nearmap committed:
COMM-455 Fix issue with scope claims handling on RefreshTokenStore
- Have fixed an issue with the way scopes claims were being handled, upgrade didn't test this correctly and values were not able to deserialise. Now fixed.
- Have simplified serialisation models to remove duplicate fields. There was quite a bit of data being serialised multiple times due to layering. This has now been streamlined.

JasonBridgman-Nearmap committed:
Merge pull request #2 from nearmap/COMM-455-fix-RefreshTokenStore-issue

COMM-542 Fix issue with scope claims handling on RefreshTokenStore
